### DISCUSSION TITLE
New ESP32 Candidate for v3!

### DISCUSSION THREAD

**Discussion by 3DSmitty** [luc-github/ESP3D#802] [General]
Hello! I just discovered this new ESP32 development board that is really inexpensive and features:

*Use low-power dual-core 32-bit ESP-WROOM-32 CPU
*The main frequency is up to 240MHz, and the computing power is up to 600 DMIPS
*Built-in 520 KB SRAM
*Support UART/SPI/I2C/PWM/ADC/DAC and other interfaces
*Support OV2640 and OV7670 cameras, built-in flash
*Support TF card
*2.8" 240*320 Smart Display Screen LCD TFT Module With Touch

https://www.aliexpress.com/item/3256804315935867.html?spm=a2g0s.8937460.0.0.4b762e0edIURZR

I think would be great for a candidate for ESP3D v3 as a all in one 3D printer controller. I understand there is some LCD development going on, how hard do you think it would be to add a lcd gui / touch functions to ESP3D? What do you think about this device?

**Comment by luc-github** [comment]
yes it is what I did using components https://github.com/luc-github/ESP3D/issues/528

it is not hard but need lot of work, design UI, design userflow, define features to display, etc...

about hardware in Aliexpress, I will buy a couple but the price is really low, I feel really strange to have is so low, it looks just boom cost, so it actually looks too cheap to be good....

**Reply by 3DSmitty** [reply]
Yes I was suspect about the price too. I have one on the way so I will report back when I get it :)

**Reply by luc-github** [reply]
I have ordered 2 , they wrote will be sent within 11 days, let's see 
I see some others but without sd card and 2.5 the price of this one...

**Reply by 3DSmitty** [reply]
Ok great! Seems to have great feedback on it so we will see....

**Comment by luc-github** [comment]
this one is one level up : more memory , bigger screen, capacitive touch, and of course more expensive...
https://www.aliexpress.com/item/1005003814428825.html?spm=a2g0o.cart.0.0.611d38da2FJxHS&mp=1

**Reply by 3DSmitty** [reply]
Yes that is nice, a little bit pricey though.

**Reply by luc-github** [reply]
received the S3 based board today and it worth the price 
Quality is there definitly, product is very well realized.

I have played with it using preinstalled sample code and it is really reactive and faster than my old protype : I am excited start devt on it 

```
ESP-ROM:esp32s3-20210327
Build:Mar 27 2021
rst:0x1 (POWERON),boot:0x8 (SPI_FAST_FLASH_BOOT)
SPIWP:0xee
mode:DIO, clock div:1
load:0x3fcd0108,len:0x17b4
load:0x403b6000,len:0xe64
load:0x403ba000,len:0x30d4
entry 0x403b6280
[0;32mI (25) boot: ESP-IDF 8c8f700c1d-dirty 2nd stage bootloader[0m
[0;32mI (25) boot: compile time 16:06:11[0m
[0;32mI (25) boot: chip revision: 0[0m
[0;32mI (27) qio_mode: Enabling QIO for flash chip WinBond[0m
[0;32mI (33) boot.esp32s3: Boot SPI Speed : 80MHz[0m
[0;32mI (38) boot.esp32s3: SPI Mode       : QIO[0m
[0;32mI (42) boot.esp32s3: SPI Flash Size : 8MB[0m
[0;32mI (47) boot: Enabling RNG early entropy source...[0m
[0;32mI (53) boot: Partition Table:[0m
[0;32mI (56) boot: ## Label            Usage          Type ST Offset   Length[0m
[0;32mI (63) boot:  0 nvs              WiFi data        01 02 00009000 00006000[0m
[0;32mI (71) boot:  1 phy_init         RF data          01 01 0000f000 00001000[0m
[0;32mI (78) boot:  2 factory          factory app      00 00 00010000 002ee000[0m
[0;32mI (86) boot: End of partition table[0m
[0;32mI (90) esp_image: segment 0: paddr=00010020 vaddr=3c060020 size=3e1a0h (254368) map[0m
[0;32mI (137) esp_image: segment 1: paddr=0004e1c8 vaddr=3fc96000 size=01e50h (  7760) load[0m
[0;32mI (139) esp_image: segment 2: paddr=00050020 vaddr=42000020 size=51da0h (335264) map[0m
[0;32mI (194) esp_image: segment 3: paddr=000a1dc8 vaddr=3fc97e50 size=00b24h (  2852) load[0m
[0;32mI (194) esp_image: segment 4: paddr=000a28f4 vaddr=40378000 size=0dff8h ( 57336) load[0m
[0;32mI (210) esp_image: segment 5: paddr=000b08f4 vaddr=50000000 size=00010h (    16) load[0m
[0;32mI (216) boot: Loaded app from partition at offset 0x10000[0m
[0;32mI (216) boot: Disabling RNG early entropy source...[0m
[0;32mI (231) cpu_start: Pro cpu up.[0m
[0;32mI (231) cpu_start: Starting app cpu, entry point is 0x403791ec[0m
[0;32mI (0) cpu_start: App cpu up.[0m
[0;32mI (245) cpu_start: Pro cpu start user code[0m
[0;32mI (245) cpu_start: cpu freq: 240000000[0m
[0;32mI (245) cpu_start: Application information:[0m
[0;32mI (248) cpu_start: Project name:     lvgl_demo[0m
[0;32mI (253) cpu_start: App version:      1[0m
[0;32mI (257) cpu_start: Compile time:     Jun 29 2022 16:05:43[0m
[0;32mI (263) cpu_start: ELF file SHA256:  55d96109010b1fb4...[0m
[0;32mI (269) cpu_start: ESP-IDF:          8c8f700c1d-dirty[0m
[0;32mI (275) heap_init: Initializing. RAM available for dynamic allocation:[0m
[0;32mI (282) heap_init: At 3FC99950 len 000466B0 (281 KiB): D/IRAM[0m
[0;32mI (288) heap_init: At 3FCE0000 len 0000EE34 (59 KiB): STACK/DRAM[0m
[0;32mI (295) heap_init: At 3FCF0000 len 00008000 (32 KiB): DRAM[0m
[0;32mI (301) heap_init: At 600FE000 len 00002000 (8 KiB): RTCRAM[0m
[0;32mI (308) spi_flash: detected chip: winbond[0m
[0;32mI (312) spi_flash: flash io: qio[0m
[0;32mI (317) sleep: Configure to isolate all GPIO pins in sleep state[0m
[0;32mI (323) sleep: Enable automatic switching of GPIO sleep configuration[0m
[0;32mI (330) cpu_start: Starting scheduler on PRO CPU.[0m
[0;32mI (0) cpu_start: Starting scheduler on APP CPU.[0m
[0;32mI (351) i2c_bus: i2c0 bus inited[0m
[0;32mI (361) tca9554: tca9554 init ok[0m
[0;32mI (361) lv_port_disp: init lvgl lcd display port[0m
[0;32mI (361) gpio: GPIO[21]| InputEn: 0| OutputEn: 1| OpenDrain: 0| Pullup: 0| Pulldown: 0| Intr:0 [0m
[0;32mI (511) rm68120: LCD=800x480 dir=0 xgap=0 ygap=0[0m
[0;32mI (511) rm68120: madctl=0x60 colmod=0x55[0m
[0;32mI (511) ft5x06: ft5x06 init ok[0m

```
I have contacted the Dev and he is very responsive :+1:

The other board is already sent but still in transit - oversea ...

**Reply by luc-github** [reply]
I have created a repo for this project https://github.com/luc-github/ESP3D-TFT
because core is TFT - WIFI come as addon, so I will use lvgl / idf instead of arduino 
It will also give me oportunity to test / use idf webserver / websocket libs - so big task is ongoing ...

**Reply by 3DSmitty** [reply]
Awesome! This could potentially turn into something very cool! I am still waiting for the small yellow one to arrive.

**Comment by 3DSmitty** [comment]
This one (The PLUS version) looks promising at a reasonable price!

https://www.aliexpress.com/item/3256804123511422.html?spm=a2g0s.8937460.0.0.2b162e0e26hgB4

**Reply by luc-github** [reply]
yes I saw this one you but I only saw the no SD version - did not noticed the plus has SD
For me price is even better 
<img width="823" alt="image" src="https://user-images.githubusercontent.com/8822552/186438817-f2fc6f64-b4a8-4d23-9a8a-b46b230b6cc4.png">

But the issue is it only has 4MB flash which won't allow bitmap to be stored on it - so will need to use sd card and the display will then be slow dow due to sd access for bitmap

**Reply by 3DSmitty** [reply]
That may be true for the regular version but the PLUS version (if you look at the pdf link) has the WT32-S3-WROVER-N8R2 ESP MCU. The WT32-S3-WROVER-N8R2 (if you google the datasheet) has 8MB flash and 2MB PSRAM built in.

**Reply by luc-github** [reply]
I only rely on vendor page who wrote : 4MB SPI Flash 8MB PSRAM  
do you see another specs in the vendor page ? may be it worse asking him confirmation for the plus

**Reply by 3DSmitty** [reply]
yeah I could ask.

**Reply by 3DSmitty** [reply]
[8a8a8a1a82380f240182af7fa91e0058.pdf](https://github.com/luc-github/ESP3D/files/9417180/8a8a8a1a82380f240182af7fa91e0058.pdf)

Attached is the vendor pdf sheet. Last page has contact information. I will reach out.

**Reply by 3DSmitty** [reply]
SO I have confirmed the regular version has the ESP32-WROVER-B which is 4MB Flash and 8MB PSRAM. The Plus version has the WT32-S3-WROVER-N8R2 which is 8MB Flash and 2MB PSRAM. I just ordered one.

**Reply by luc-github** [reply]
OK thank you for the information - I have ordered 1 too - I think it cover all resolutions now

**Comment by luc-github** [comment]
Just found picture of my original prototype 4 years ago
https://github.com/luc-github/ESP3D/issues/239#issuecomment-459045357

Now I have better boards but my design skills still did not improved in 4 years haha - I hope I can rely on lvgl common controls for the UI as they become better since then

**Reply by 3DSmitty** [reply]
Awesome!

**Comment by 3DSmitty** [comment]
So the original little yellow lcd board came in. Looks pretty cool for the price. It has the ESP32-D0WDQ6 MCU with no flash memory onboard. Attached is a Winbond SPI flash-chip with 32M-bit or 4M-byte. Do you think this is enough space for ESP3D and minimal printer control graphics?

**Reply by luc-github** [reply]
well everything depend of the UI used for reference this is the bitmaps used for BTT TFT 2.8 and 3.5
4MB for Firmware + update + bitmap is littlebit tight

For info a  classic ESP3D FW with wifi + BT + update only left 190KB  for flash for web server (min_spiffs.csv) so  where to put the pictures for the UI ? on SD the UI may be very slow but need to be tested

![image](https://user-images.githubusercontent.com/8822552/187032252-a95b53c1-022b-4edd-b2f7-07f5938d0293.png)

**Reply by 3DSmitty** [reply]
I have done many bios chip / SPI flash chip reworks on PC motherboards in my previous life. Looking at the board there is really nothing in the way to do so, Easy Peasy. I am going to research a larger compatible chip. It would even be nice to put a socket.

**Reply by 3DSmitty** [reply]
What about this one:
Winbond W25Q128JV
https://www.winbond.com/hq/support/documentation/levelOne.jsp?__locale=en&DocNo=DA00-W25Q128JV

This would give us 16M-byte to play with. It is the same thing just with larger memory.

**Reply by 3DSmitty** [reply]
Just ordered a 5-pack of the Winbond W25Q128JVSIQ

**Reply by luc-github** [reply]
Ho I did reworked an esp8266 1M to do 4M this was not easy for my fat fingers...
thanks for the info

**Comment by Skorpi08** [comment]
https://www.youtube.com/watch?v=WgIDNHImL7A&ab_channel=Skorpi

SD vs SPI Flash

**Reply by 3DSmitty** [reply]
@Skorpi08 yeah you can clearly see the difference, hmm.

**Reply by luc-github** [reply]
That is why I wrote before design UI won't be an easy task ^_^

**Reply by luc-github** [reply]
@3DSmitty I have started to collect informations about references screens https://github.com/luc-github/ESP3D-TFT/tree/main/references
If you want to contribute please feel free to add page / docs on the TFT repo about the screen(s) you get

**Reply by 3DSmitty** [reply]
@luc-github  Absolutely, thank you!

**Reply by 3DSmitty** [reply]
@luc-github I am going to start wrapping my head around how the puzzle fits together. Should I start by getting ESP3D v3.0 running on the device and then implementing lvgl? How do the buttons on the WEBUI send commands to Marlin? Are they hardcoded or are they library variables? If library variables then it should be easy to create TFT buttons that do the same thing. Also what is wrong with littlefs?

**Reply by luc-github** [reply]
there is  0 support in esp3d for tft UI and buttons, every commands are coded in webui
adding a  lvgl button is not that trivial fyi
the performance for a tft interface in esp3d using arduino are not great that is why better go to idf
 what is your issue with littlefs?

**Reply by 3DSmitty** [reply]
I was thinking of using a platformio project with the espressif libraries instead of the arduino libraries. Would it make sense to create a library of variables or a header file  with (Marlin / GRBL Commands) that either the WEBUI or TFT or both references?
Yes I have played with lvgl a little. Creating buttons are not hard.
I have no issue with littleFS at all. You put on the ESP3D-TFT github

> Unlike ESP3D some features will be fixed:
>
> * Flash FS = littleFS
> * TBD

**Reply by luc-github** [reply]
My meaning was fixed vs variable  (ESP3D support SPIFFS/FAT/LittleFS) the tft should only need/use LittleFS

No it is not hard to do button but  
1 - need TFT drivers (SPI / IC2/ Pins?)
2 - need touche drivers (SPI / IC2 / Pins?)
3 - need a UI flow for this button  (TBD)
4 - need UI images corresponding to resolution (280x320) (480X320) (800x600)
That is why I said it is not trivial - there are several prerequisits to consider and do

The UI between 3D Printer and CNC are totaly different - I am not sure to understand what you mean by create a library of variables

Did you checked the repository I have created for the ESP3D-TFT ? there is a Status with possible ways to achieve

**Reply by 3DSmitty** [reply]
Sorry I think I may have explained it wrong. I mean a header file that both WEBUI and TFT could share that included Marlin / GRBL commands if that makes sense?

**Reply by Skorpi08** [reply]
Luc, please also think of FluidNC ;)

**Reply by luc-github** [reply]
WebUI is using Javascript and commands are modifiable by User there is no `header file` but JSON
TFT will use C and may be C++ and the commands can also be modifiable by some ini file
Both could share same concept but not same code

**Reply by luc-github** [reply]
@Skorpi08 FluidNC specificaly is out of my area now - they asked me to not work on it when I started to add support on WebUI 3.0 for FluidNC, and they do they own fork since a while, but as GRBL variant, FluidNC should work as the CNC UI would support any GRBL based FW 
But to be honest I do not see what whould be the benefit on FluidNC, which is already supposed support WiFi , to connect a serial wifi tft 
Do you have user case ?

**Reply by 3DSmitty** [reply]
@luc-github gotcha, thanks for explaining, that makes total sense. Oh I just noticed the discussion on the ESP3D-TFT repo :)

**Reply by luc-github** [reply]
There is big difference between ESP3D and ESP3D-TFT
ESP3D is a wifi bridge only and currently arch is based on this idea, adding screen is just for information
ESP3D-TFT is a TFT first and additionnaly a wifi bridge, so arch need to be reviewed
At the end I hope can have one code only or same core  for all (like ESP3D V3 aqnd ESP3DLib V3 ) but currently I think focus on TFT performances / features implementation  is the  key point
